File: tools/opcount.py
#!/usr/bin/env python3
"""Count guest instructions executed by a decode, via a QEMU TCG plugin.

    python3 tools/opcount.py --target riscv32 vectors/l3-hecommon.bit

This is the repository's cross-target performance metric, and the one nothing
else here provides. Static .text bounds a win but cannot rank changes. The
host's Callgrind counts x86-64 instructions for a decoder that ships to 32-bit
targets, and on this decoder it has been wrong about both the sign and the
magnitude. Wall clock under qemu-user measures the JIT.

An executed instruction count is exact, reproducible bit for bit, and free of
timing noise: two runs of the same binary on the same input return the same
number on any machine. That is a real advantage over a throughput benchmark,
and it is why this repository can gate on it rather than merely report it.

It is not a cycle count. It does not know the target's pipeline, cache or
multiplier latency, so it ranks candidates rather than predicting hardware
time. Real silicon remains the timing authority (FastLED's `bash mp3measure`).
"""

from __future__ import annotations

import argparse
import os
import re
import shutil
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
BUILD = ROOT / ".build"
PLUGIN_SRC = ROOT / "tools" / "qemu-plugin" / "opcount.c"

TARGETS = {
    "riscv32": ("riscv32-linux-musl", "qemu-riscv32"),
    "arm": ("arm-linux-musleabihf", "qemu-arm"),
}

OPCOUNT_RE = re.compile(r"OPCOUNT (\d+)")


def qemu_include_dir() -> Path | None:
    """Where qemu-plugin.h lives. It ships inside the QEMU package rather than
    a separate -dev, so it is found relative to the binary on the PATH.
    MINIMP3_QEMU_INCLUDE overrides, which is how CI points at a header fetched
    to match the distribution's QEMU -- Debian and Ubuntu package the emulator
    without it."""
    override = os.environ.get("MINIMP3_QEMU_INCLUDE")
    if override and (Path(override) / "qemu-plugin.h").is_file():
        return Path(override)
    binary = shutil.which("qemu-riscv32") or shutil.which("qemu-arm")
    if not binary:
        return None
    for candidate in (Path(binary).resolve().parent.parent / "include",
                      Path("/usr/include")):
        if (candidate / "qemu-plugin.h").is_file():
            return candidate
    return None


def build_plugin() -> Path | None:
    """Compile the counter. Returns None when the toolchain cannot -- callers
    then report the metric as unavailable rather than reporting a wrong one."""
    BUILD.mkdir(exist_ok=True)
    plugin = BUILD / "libopcount.so"
    include = qemu_include_dir()
    if include is None:
        return None
    cflags: list[str] = []
    if shutil.which("pkg-config"):
        probe = subprocess.run(["pkg-config", "--cflags", "glib-2.0"],
                               capture_output=True, text=True)
        if probe.returncode == 0:
            cflags = probe.stdout.split()
    if not cflags:
        return None  # qemu-plugin.h includes glib.h; without it, no plugin.
    result = subprocess.run(
        ["cc", "-shared", "-fPIC", "-O2", f"-I{include}", *cflags,
         str(PLUGIN_SRC), "-o", str(plugin)],
        capture_output=True, text=True)
    return plugin if result.returncode == 0 else None


def build_decoder(target: str, fixed: bool = True) -> Path:
    BUILD.mkdir(exist_ok=True)
    triple, _ = TARGETS[target]
    variant = "fixed" if fixed else "float"
    binary = BUILD / f"decode_{variant}_{target}"
    subprocess.run(
        ["zig", "c++", "-target", triple, "-static", "-nostdlib++",
         "-std=gnu++11", "-Os", "-fno-exceptions", "-fno-rtti",
         "-fno-strict-aliasing", "-DMINIMP3_NO_SIMD",
         f"-DMINIMP3_{'FIXED' if fixed else 'FLOAT'}_POINT",
         f"-I{ROOT}", f"-I{ROOT / 'port'}",
         str(ROOT / "tools" / "decode.cpp"), "-o", str(binary)],
        cwd=ROOT, check=True, capture_output=True)
    return binary


def count(plugin: Path, target: str, binary: Path, bitstream: Path) -> int | None:
    _, runner = TARGETS[target]
    result = subprocess.run(
        [runner, "-plugin", str(plugin), str(binary), str(bitstream)],
        capture_output=True, text=True)
    match = OPCOUNT_RE.search(result.stderr)
    return int(match.group(1)) if match else None


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("bitstreams", nargs="*", type=Path)
    parser.add_argument("--target", default="riscv32", choices=sorted(TARGETS))
    parser.add_argument("--float", action="store_true", dest="float_variant")
    args = parser.parse_args(argv)

    plugin = build_plugin()
    if plugin is None:
        print("opcount unavailable: needs QEMU with plugin support, its "
              "qemu-plugin.h, and glib development headers.", file=sys.stderr)
        return 2
    binary = build_decoder(args.target, fixed=not args.float_variant)
    streams = args.bitstreams or sorted((ROOT / "vectors").glob("*.bit"))
    total = 0
    for bitstream in streams:
        n = count(plugin, args.target, binary, bitstream)
        if n is None:
            print(f"  {bitstream.stem:40s} FAILED")
            continue
        total += n
        if len(streams) <= 12:
            print(f"  {bitstream.stem:40s} {n:>14,}")
    print(f"  {'TOTAL (' + args.target + ')':40s} {total:>14,}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/qemu-plugin/opcount.c
/* A QEMU TCG plugin that counts guest instructions executed.
 *
 * This is the repository's deterministic cross-target performance metric. It
 * answers "how many riscv32 instructions does a decode actually execute",
 * which is the number that matters and the one nothing else here provides:
 * static .text bounds a win but cannot rank changes, the host's Callgrind
 * counts x86-64 instructions for a decoder that ships to 32-bit targets, and
 * wall clock under qemu-user measures the JIT rather than the guest.
 *
 * It is exact and reproducible: the same binary on the same input returns the
 * same count every time, on any host, with no timing noise at all. It is not
 * a cycle count -- it does not know about the target's pipeline, its cache, or
 * its multiplier latency -- so it ranks candidates rather than predicting
 * hardware time. Real silicon stays the timing authority.
 *
 * Build with tools/opcount.py, which supplies the include path and the
 * QEMU-version-matched header.
 */

#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>

#include <qemu-plugin.h>

QEMU_PLUGIN_EXPORT int qemu_plugin_version = QEMU_PLUGIN_VERSION;

static struct qemu_plugin_scoreboard *counts;
static qemu_plugin_u64 insn_count;

static void vcpu_tb_trans(qemu_plugin_id_t id, struct qemu_plugin_tb *tb)
{
    size_t n = qemu_plugin_tb_n_insns(tb);

    /* One accumulate per translation block rather than per instruction: the
       block's instruction count is known at translation time, so the executed
       total is exact while costing a single add per block at run time. */
    qemu_plugin_register_vcpu_tb_exec_inline_per_vcpu(
        tb, QEMU_PLUGIN_INLINE_ADD_U64, insn_count, n);
}

static void plugin_exit(qemu_plugin_id_t id, void *p)
{
    uint64_t total = qemu_plugin_u64_sum(insn_count);
    /* stderr so it cannot be confused with the decoder's own stdout, which
       carries the checksum and PSNR the harness parses. */
    fprintf(stderr, "OPCOUNT %" PRIu64 "\n", total);
}

QEMU_PLUGIN_EXPORT int qemu_plugin_install(qemu_plugin_id_t id,
                                           const qemu_info_t *info,
                                           int argc, char **argv)
{
    counts = qemu_plugin_scoreboard_new(sizeof(uint64_t));
    insn_count = qemu_plugin_scoreboard_u64(counts);
    qemu_plugin_register_vcpu_tb_trans_cb(id, vcpu_tb_trans);
    qemu_plugin_register_atexit_cb(id, plugin_exit, NULL);
    return 0;
}
